converter/src/main/java/gov/cms/qpp/conversion/decode/QrdaDecoderEngine.java

```diff
@@ -0,0 +1,284 @@
+	private static final String NOT_VALID_QRDA_III_FORMAT = "The file is not a QRDA-III XML document";
+	private static final String ROOT_STRING = "root";
+	private static final String EXTENSION_STRING = "extension";
+
+	protected final Context context;
+	private final Set<TemplateId> scope;
+	private final Registry<QrdaDecoder> decoders;
+
+	/**
+	 * Initialize a QPP xml decoder
+	 */
+	public QrdaDecoderEngine(Context context) {
+		Objects.requireNonNull(context, "converter");
+
+		this.context = context;
+		this.scope = context.hasScope() ? QrdaScope.getTemplates(context.getScope()) : null;
+		this.decoders = context.getRegistry(Decoder.class);
+	}
+
+	/**
+	 * Decodes the top of the XML document.
+	 *
+	 * @param xmlDoc XML Document to be parsed.
+	 * @return The root node.
+	 */
+	@Override
+	public Node decode(Element xmlDoc) {
+		Node rootNode = new Node();
+		Element rootElement = xmlDoc.getDocument().getRootElement();
+		defaultNs = rootElement.getNamespace();
+
+		rootNode.setType(TemplateId.PLACEHOLDER);
+		rootNode.setPath(XPathHelper.getAbsolutePath(rootElement));
+
+		QrdaDecoder rootDecoder = null;
+		for (Element element : rootElement.getChildren(TEMPLATE_ID, rootElement.getNamespace())) {
+			rootDecoder = getDecoder(getTemplateId(element));
+			if (rootDecoder != null) {
+				break;
+			}
+		}
+
+		if (rootDecoder != null) {
+			rootNode = this.decodeTree(rootElement, rootNode).getNode().getChildNodes().get(0);
+		} else {
+			rootNode = this.decodeTree(rootElement, rootNode).getNode();
+		}
+
+		return rootNode;
+	}
+
+	/**
+	 * Decodes the element specified and the entire tree of child {@link Element}s below.
+	 *
+	 * @param element The element who's tree to decode.
+	 * @param parentNode The node to add any possible decoded child {@link Node}s.
+	 * @return The tuple of a {@link DecodeResult} and {@link Node} that was decoded from this tree.
+	 */
+	private DecodeData decodeTree(final Element element, final Node parentNode) {
+		DecodeData result = decodeSingleElement(element, parentNode);
+		DecodeResult decodedResult = result.getDecodeResult();
+		Node decodedNode = result.getNode();
+
+		decodedNode = decideNewParentNode(decodedNode, parentNode);
+
+		if (DecodeResult.TREE_FINISHED == decodedResult) {
+			return new DecodeData(DecodeResult.TREE_FINISHED, decodedNode);
+		} else if (DecodeResult.TREE_ESCAPED == decodedResult) {
+			return new DecodeData(DecodeResult.TREE_FINISHED, null);
+		}
+
+		return decodeChildren(element, decodedNode);
+	}
+
+	/**
+	 * Decodes the passed in element if it is a {@code templateId} and assigns it to the {@code parentNode}.
+	 *
+	 * @param element The element to decode.
+	 * @param parentNode The node add the child decoded {@link Node} to.
+	 * @return The tuple of a {@link DecodeResult} and {@link Node} that was decoded from the {@link Element}.
+	 */
+	private DecodeData decodeSingleElement(final Element element, final Node parentNode) {
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Reduce the number of returns of this method 4, down to the maximum allowed 3. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1142)